<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ malan</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
<meta property="og:title" content="Function reference" />


<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->




  </head>

  <body>
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">malan</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">1.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/case-study-partial.html">Case study: Partial profiles</a>
    </li>
    <li>
      <a href="../articles/introduction.html">Introduction</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        <!--  -->
        <td>
          <p><code><a href="analyse_mixture_result.html">analyse_mixture_result()</a></code> </p>
        </td>
        <td><p>Analyse mixture results</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="analyse_mixture_results.html">analyse_mixture_results()</a></code> </p>
        </td>
        <td><p>Analyse mixture results in a vectorised fashion</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="as_tbl_graph.malan_pedigreelist.html">as_tbl_graph(<i>&lt;malan_pedigreelist&gt;</i>)</a></code> </p>
        </td>
        <td><p>Get tidy graph object</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="brothers_matching.html">brothers_matching()</a></code> </p>
        </td>
        <td><p>Number of brothers with matching haplotype</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="build_pedigrees.html">build_pedigrees()</a></code> </p>
        </td>
        <td><p>Build pedigrees from (individuals in) a population.</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="calc_autosomal_genotype_conditional_cumdist.html">calc_autosomal_genotype_conditional_cumdist()</a></code> </p>
        </td>
        <td><p>Calculate conditional genotype cumulative probabilities with theta</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="calc_autosomal_genotype_probs.html">calc_autosomal_genotype_probs()</a></code> </p>
        </td>
        <td><p>Calculate genotype probabilities with theta</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="count_brothers.html">count_brothers()</a></code> </p>
        </td>
        <td><p>Number of brothers</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="count_haplotype_near_matches_individuals.html">count_haplotype_near_matches_individuals()</a></code> </p>
        </td>
        <td><p>Count near haplotype matches in list of individuals</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="count_haplotype_occurrences_individuals.html">count_haplotype_occurrences_individuals()</a></code> </p>
        </td>
        <td><p>Count haplotypes occurrences in list of individuals</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="count_haplotype_occurrences_pedigree.html">count_haplotype_occurrences_pedigree()</a></code> </p>
        </td>
        <td><p>Count haplotypes occurrences in pedigree</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="count_uncles.html">count_uncles()</a></code> </p>
        </td>
        <td><p>Number of uncles</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="estimate_autotheta_1subpop_genotypes.html">estimate_autotheta_1subpop_genotypes()</a></code> </p>
        </td>
        <td><p>Estimate autosomal theta from genotypes</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="estimate_autotheta_1subpop_individuals.html">estimate_autotheta_1subpop_individuals()</a></code> </p>
        </td>
        <td><p>Estimate autosomal theta from individuals</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="estimate_autotheta_subpops_genotypes.html">estimate_autotheta_subpops_genotypes()</a></code> </p>
        </td>
        <td><p>Estimate autosomal F, theta, and f from subpopulations of genotypes</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="estimate_autotheta_subpops_individuals.html">estimate_autotheta_subpops_individuals()</a></code> </p>
        </td>
        <td><p>Estimate autosomal F, theta, and f from subpopulations of individuals</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="estimate_autotheta_subpops_pids.html">estimate_autotheta_subpops_pids()</a></code> </p>
        </td>
        <td><p>Estimate autosomal F, theta, and f from subpopulations of individual ids</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="estimate_autotheta_subpops_unweighted_genotypes.html">estimate_autotheta_subpops_unweighted_genotypes()</a></code> </p>
        </td>
        <td><p>Unweighted estimate of autosomal theta from subpopulations of genotypes</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="estimate_autotheta_subpops_unweighted_pids.html">estimate_autotheta_subpops_unweighted_pids()</a></code> </p>
        </td>
        <td><p>Unweighted estimate of autosomal theta from subpopulations of individual ids</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="father_matches.html">father_matches()</a></code> </p>
        </td>
        <td><p>Father matches</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="generate_get_founder_haplotype_db.html">generate_get_founder_haplotype_db()</a></code> </p>
        </td>
        <td><p>Generate a function to simulate pedigree founder haplotype based on a haplotype databasep</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="generate_get_founder_haplotype_ladder.html">generate_get_founder_haplotype_ladder()</a></code> </p>
        </td>
        <td><p>Generate a function to simulate pedigree founder haplotype based on ladder information</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="get_allele_counts_genotypes.html">get_allele_counts_genotypes()</a></code> </p>
        </td>
        <td><p>Get autosomal allele counts from subpopulations of genotypes</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="get_allele_counts_pids.html">get_allele_counts_pids()</a></code> </p>
        </td>
        <td><p>Get autosomal allele counts from subpopulations given by pids</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="get_brothers.html">get_brothers()</a></code> </p>
        </td>
        <td><p>Get brothers</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="get_children.html">get_children()</a></code> </p>
        </td>
        <td><p>Get children</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="get_cousins.html">get_cousins()</a></code> </p>
        </td>
        <td><p>Get cousins</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="get_family_info.html">get_family_info()</a></code> </p>
        </td>
        <td><p>Get individual's family information</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="get_generation.html">get_generation()</a></code> </p>
        </td>
        <td><p>Get individual's generation number</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="get_haplotype.html">get_haplotype()</a></code> </p>
        </td>
        <td><p>Get haplotype from an individual</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="get_haplotypes_in_pedigree.html">get_haplotypes_in_pedigree()</a></code> </p>
        </td>
        <td><p>Get haplotypes in pedigree</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="get_haplotypes_individuals.html">get_haplotypes_individuals()</a></code> </p>
        </td>
        <td><p>Get haplotype matrix from list of individuals</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="get_haplotypes_pids.html">get_haplotypes_pids()</a></code> </p>
        </td>
        <td><p>Get haplotypes from a vector of pids.</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="get_individual.html">get_individual()</a></code> </p>
        </td>
        <td><p>Get individual by pid</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="get_individuals.html">get_individuals()</a></code> </p>
        </td>
        <td><p>Get all individuals in population</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="get_nodes_edges.html">get_nodes_edges()</a></code> </p>
        </td>
        <td><p>Get nodes and edges</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="get_pedigree_as_graph.html">get_pedigree_as_graph()</a></code> </p>
        </td>
        <td><p>Get pedigree information as graph (mainly intended for plotting)</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="get_pedigree_from_individual.html">get_pedigree_from_individual()</a></code> </p>
        </td>
        <td><p>Get pedigree from individual</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="get_pedigree_id.html">get_pedigree_id()</a></code> </p>
        </td>
        <td><p>Get pedigree id</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="get_pedigree_id_from_pid.html">get_pedigree_id_from_pid()</a></code> </p>
        </td>
        <td><p>Get pedigree ids from pids</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="get_pedigrees_tidy.html">get_pedigrees_tidy()</a></code> </p>
        </td>
        <td><p>Get pedigrees information in tidy format</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="get_pid.html">get_pid()</a></code> </p>
        </td>
        <td><p>Get pid from individual</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="get_pids_in_pedigree.html">get_pids_in_pedigree()</a></code> </p>
        </td>
        <td><p>Get pids in pedigree</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="get_uncles.html">get_uncles()</a></code> </p>
        </td>
        <td><p>Get uncles</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="grandfather_matches.html">grandfather_matches()</a></code> </p>
        </td>
        <td><p>Grandfather matches</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="haplotype_matches_individuals.html">haplotype_matches_individuals()</a></code> </p>
        </td>
        <td><p>Get individuals matching from list of individuals</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="haplotype_partially_matches_individuals.html">haplotype_partially_matches_individuals()</a></code> </p>
        </td>
        <td><p>Get individuals partially matching from list of individuals</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="haplotypes_to_hashes.html">haplotypes_to_hashes()</a></code> </p>
        </td>
        <td><p>Convert haplotypes to hashes (integers)</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="malan-package.html">malan-package</a></code> </p>
        </td>
        <td><p>MAle Lineage ANalysis</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="meioses_generation_distribution.html">meioses_generation_distribution()</a></code> </p>
        </td>
        <td><p>Meiotic distribution</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="meiotic_dist.html">meiotic_dist()</a></code> </p>
        </td>
        <td><p>Meiotic distance between two individuals</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="mixture_info_by_individuals_2pers.html">mixture_info_by_individuals_2pers()</a></code> </p>
        </td>
        <td><p>Mixture information about 2 persons' mixture of donor1 and donor2.</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="mixture_info_by_individuals_3pers.html">mixture_info_by_individuals_3pers()</a></code> </p>
        </td>
        <td><p>Mixture information about 3 persons' mixture of donor1, donor2 and donor3.</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="mixture_info_by_individuals_4pers.html">mixture_info_by_individuals_4pers()</a></code> </p>
        </td>
        <td><p>Mixture information about 4 persons' mixture of donor1, donor2, donor3 and donor4.</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="mixture_info_by_individuals_5pers.html">mixture_info_by_individuals_5pers()</a></code> </p>
        </td>
        <td><p>Mixture information about 5 persons' mixture of donor1, donor2, donor3, donor4 and donor5.</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="pedigree_as_igraph.html">pedigree_as_igraph()</a></code> </p>
        </td>
        <td><p>Convert pedigree to igraph</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="pedigree_haplotype_matches_in_pedigree_meiosis_L1_dists.html">pedigree_haplotype_matches_in_pedigree_meiosis_L1_dists()</a></code> </p>
        </td>
        <td><p>Information about matching individuals</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="pedigree_haplotype_near_matches_meiosis.html">pedigree_haplotype_near_matches_meiosis()</a></code> </p>
        </td>
        <td><p>Information about almost matching individuals</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="pedigree_size.html">pedigree_size()</a></code> </p>
        </td>
        <td><p>Get pedigree size</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="pedigree_size_generation.html">pedigree_size_generation()</a></code> </p>
        </td>
        <td><p>Size of pedigree</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="pedigrees_all_populate_autosomal.html">pedigrees_all_populate_autosomal()</a></code> </p>
        </td>
        <td><p>Populate 1-locus autosomal DNA profile in pedigrees with single-step mutation model.</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="pedigrees_all_populate_haplotypes.html">pedigrees_all_populate_haplotypes()</a></code> </p>
        </td>
        <td><p>Populate haplotypes in pedigrees (0-founder/unbounded).</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="pedigrees_all_populate_haplotypes_custom_founders.html">pedigrees_all_populate_haplotypes_custom_founders()</a></code> </p>
        </td>
        <td><p>Populate haplotypes in pedigrees (custom founder/unbounded).</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="pedigrees_all_populate_haplotypes_ladder_bounded.html">pedigrees_all_populate_haplotypes_ladder_bounded()</a></code> </p>
        </td>
        <td><p>Populate haplotypes in pedigrees (custom founder/bounded).</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="pedigrees_count.html">pedigrees_count()</a></code> </p>
        </td>
        <td><p>Get number of pedigrees</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="pedigrees_table.html">pedigrees_table()</a></code> </p>
        </td>
        <td><p>Get distribution of pedigree sizes</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="plot.malan_pedigree.html">plot(<i>&lt;malan_pedigree&gt;</i>)</a></code> </p>
        </td>
        <td><p>Plot pedigree</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="plot.malan_pedigreelist.html">plot(<i>&lt;malan_pedigreelist&gt;</i>)</a></code> </p>
        </td>
        <td><p>Plot pedigree list</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="population_populate_autosomal_infinite_alleles.html">population_populate_autosomal_infinite_alleles()</a></code> </p>
        </td>
        <td><p>Populate 1-locus autosomal DNA profile in pedigrees with infinite alleles mutation model.</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="population_size_generation.html">population_size_generation()</a></code> </p>
        </td>
        <td><p>Size of population</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="print.malan_pedigree.html">print(<i>&lt;malan_pedigree&gt;</i>)</a></code> </p>
        </td>
        <td><p>Print pedigree</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="print.malan_pedigreelist.html">print(<i>&lt;malan_pedigreelist&gt;</i>)</a></code> </p>
        </td>
        <td><p>Print pedigree list</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="print.malan_population.html">print(<i>&lt;malan_population&gt;</i>)</a></code> </p>
        </td>
        <td><p>Print population</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="print.malan_population_abort.html">print(<i>&lt;malan_population_abort&gt;</i>)</a></code> </p>
        </td>
        <td><p>Print <code>malan_population_abort</code></p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="print_individual.html">print_individual()</a></code> </p>
        </td>
        <td><p>Print individual</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="sample_autosomal_genotype.html">sample_autosomal_genotype()</a></code> </p>
        </td>
        <td><p>Sample genotype with theta</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="sample_geneology.html">sample_geneology()</a></code> </p>
        </td>
        <td><p>Simulate a geneology with constant population size.</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="sample_geneology_varying_size.html">sample_geneology_varying_size()</a></code> </p>
        </td>
        <td><p>Simulate a geneology with varying population size.</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="split_by_haplotypes.html">split_by_haplotypes()</a></code> </p>
        </td>
        <td><p>Split pids by haplotype</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="sub-sub-.malan_pedigreelist.html">`[[`(<i>&lt;malan_pedigreelist&gt;</i>)</a></code> </p>
        </td>
        <td><p>Get pedigree from pedigree list</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="sub-sub-.malan_population.html">`[[`(<i>&lt;malan_population&gt;</i>)</a></code> </p>
        </td>
        <td><p>Get individual from population by pid</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="test_create_population.html">test_create_population()</a></code> </p>
        </td>
        <td><p>Generate test population</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="ystr_kits.html">ystr_kits</a></code> </p>
        </td>
        <td><p>Kit information about Y-STR markers</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="ystr_markers.html">ystr_markers</a></code> </p>
        </td>
        <td><p>Mutational information about Y-STR markers</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#section-all-functions">All functions</a></li>
    </ul>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Mikkel Meyer Andersen.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

